semantic/lookup.go

```diff
@@ -30,70 +30,44 @@ func LookupBuiltInType(pkg, name string) (MonoType, error) {
 	return monotype, nil
 }
 
-// lookup is a helper function that performs a lookup of a package identifier in
-// a flatbuffers TypeEnvironment. It first checks for the package using path string
-// and then checks for the package identifier, the "name" string in this case, returning
-// the corresponding MonoType if found
-func lookup(env *fbsemantic.TypeEnvironment, pkg, name string) (*fbsemantic.Prop, error) {
-	// Find package
-	typeAssign := new(fbsemantic.TypeAssignment)
-	if pkgErr := foundPackage(env, typeAssign, pkg); pkgErr != nil {
-		return nil, pkgErr
-	}
-
-	// Grab PolyType expr; expr should always be of type MonoTypeRow
-	polytype := typeAssign.Ty(nil)
-	if polytype.ExprType() != fbsemantic.MonoTypeRow {
-		return nil, errors.Newf(
-			codes.Internal,
-			"Expected PolyType expr to be fbsemantic.MonoTypeRow; found %v",
-			polytype.ExprType(),
-		)
-	}
-	var table flatbuffers.Table
-	if !polytype.Expr(&table) {
-		return nil, errors.New(codes.Internal, "PolyType expr is not valid")
-	}
-
-	// check for package identifier in Row Props
-	row := new(fbsemantic.Row)
-	row.Init(table.Bytes, table.Pos)
-	prop := new(fbsemantic.Prop)
-	if propErr := foundProp(row, prop, name); propErr != nil {
-		return nil, propErr
+// MustLookupBuiltinType validates that call to LookupBuiltInType was
+// successful. If there is an error with lookup, then panic.
+func MustLookupBuiltinType(pkg, name string) MonoType {
+	mt, err := LookupBuiltinType(pkg, name)
+	if err != nil {
+		panic(err)
 	}
-
-	return prop, nil
+	return mt
 }
 
-// foundPackage is a helper function that iterates over type assignments and checks
-// for a given package returning an error if the package is not found
-func foundPackage(env *fbsemantic.TypeEnvironment, obj *fbsemantic.TypeAssignment, pkg string) error {
+// TypeEnvMap creates a global map of the TypeEnvironment
+func TypeEnvMap(env *fbsemantic.TypeEnvironment) map[string]map[string]*fbsemantic.Prop {
+	envMap := make(map[string]map[string]*fbsemantic.Prop)
```

Comment:
Go lets you use a struct as the key to a map, in most cases using a struct is more efficient than nested maps.

So you could create a struct like this :

```
type envKey struct {
    Package string
    Prop string
}
```

And a map like this:

```
envMap := make(map[envKey]*fbsemantic.Prop)
```


Then you can look it up like this:

```
key := envKey {
    Package: pkg,
    Prop: name,
}

prop := envMap[key]
```

This change isn't necessary but it will be more idiomatic Go.

Reply:
Yeah, that looks much more Go-like. I'll go ahead and make it look more like that

Reply:
^^ Great suggestion. I feel like the code looks a lot better. I was writing it like it was Python.